server/src/main/java/org/opensearch/repositories/blobstore/BlobStoreRepository.java

```diff
@@ -1213,57 +1281,64 @@ protected void releaseRemoteStoreLockAndCleanup(
-            threadPool.executor(ThreadPool.Names.SNAPSHOT).execute(ActionRunnable.wrap(listener, l -> {
-                try {
-                    // filtering files for which remote store lock release and cleanup succeeded,
-                    // remaining files for which it failed will be retried in next snapshot delete run.
-                    List<String> eligibleFilesToDelete = new ArrayList<>();
-                    for (String fileToDelete : filesToDelete) {
-                        if (fileToDelete.contains(SHALLOW_SNAPSHOT_PREFIX)) {
-                            String[] fileToDeletePath = fileToDelete.split("/");
-                            String indexId = fileToDeletePath[1];
-                            String shardId = fileToDeletePath[2];
-                            String shallowSnapBlob = fileToDeletePath[3];
-                            String snapshotUUID = extractShallowSnapshotUUID(shallowSnapBlob).orElseThrow();
-                            BlobContainer shardContainer = blobStore().blobContainer(indicesPath().add(indexId).add(shardId));
-                            try {
-                                releaseRemoteStoreLockAndCleanup(shardId, snapshotUUID, shardContainer, remoteStoreLockManagerFactory);
-                                eligibleFilesToDelete.add(fileToDelete);
-                            } catch (Exception e) {
-                                logger.error(
-                                    "Failed to release lock or cleanup shard for indexID {}, shardID {} " + "and snapshot {}",
-                                    indexId,
-                                    shardId,
-                                    snapshotUUID
-                                );
-                            }
-                        } else {
-                            eligibleFilesToDelete.add(fileToDelete);
+        List<Tuple<BlobPath, String>> filesToDelete = staleFilesToDeleteInBatch.poll(0L, TimeUnit.MILLISECONDS);
+        if (filesToDelete == null) {
+            return;
+        }
+        threadPool.executor(ThreadPool.Names.SNAPSHOT).execute(ActionRunnable.wrap(listener, l -> {
+            try {
+                // filtering files for which remote store lock release and cleanup succeeded,
+                // remaining files for which it failed will be retried in next snapshot delete run.
+                List<String> eligibleFilesToDelete = new ArrayList<>();
+                for (Tuple<BlobPath, String> fileToDelete : filesToDelete) {
+                    BlobPath blobPath = fileToDelete.v1();
+                    String blobName = fileToDelete.v2();
+                    boolean deleteBlob = false;
+                    if (blobName.startsWith(SHALLOW_SNAPSHOT_PREFIX)) {
+                        String snapshotUUID = extractShallowSnapshotUUID(blobName).orElseThrow();
+                        String[] parts = blobPath.toArray();
+                        int partLength = parts.length;
```

Comment:
We don't have any fixed value for the partLength. It depends on the base path supplied during repo creation. There is only 1 guarantee that it would be greater than or equal to 2. And this is anyways tested out in next 2 lines. If it is untrue, it would lead to ArrayIndexOutOfBoundsException.

Reply:
Lets assert `>=3` for now .